src/interrupt/mod.rs

```diff
@@ -0,0 +1,142 @@
+//! MSI [(Message Signaled Interrupt)](https://en.wikipedia.org/wiki/Message_Signaled_Interrupts)
+//! is another commonly used alternative in-band method of signaling an interrupt, using special
+//! in-band messages to replace traditional out-of-band assertion of dedicated interrupt lines.
+//! While more complex to implement in a device, message signaled interrupts have some significant
+//! advantages over pin-based out-of-band interrupt signaling. Message signaled interrupts are
+//! supported in PCI bus since its version 2.2, and in later available PCI Express bus. Some non-PCI
+//! architectures also use message signaled interrupts.
+//!
+//! While IRQ is a term commonly used by Operating Systems when dealing with hardware
+//! interrupts, the IRQ numbers managed by OSes are independent of the ones managed by VMM.
+//! For simplicity sake, the term `Interrupt Source` is used instead of IRQ to represent both pin-based
+//! interrupts and MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * The VMM creates an interrupt manager
+//! * The VMM creates a device manager, passing on an reference to the interrupt manager
+//! * The device manager passes on an reference to the interrupt manager to all registered devices
+//! * The guest kernel loads drivers for virtual devices
+//! * The guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * The virtual device backend requests the interrupt manager to create an interrupt group
+//!   according to guest configuration information
+
+use std::fs::File;
+use std::sync::Arc;
+
+/// Reuse std::io::Result to simplify interoperability among crates.
+pub type Result<T> = std::io::Result<T>;
+
+/// Data type to store an interrupt source identifier.
+pub type InterruptIndex = u32;
+
+/// Data type to store an interrupt source type.
+pub type InterruptType = u32;
+
+pub const PIN_IRQ: InterruptType = 0;
+pub const PCI_MSI_IRQ: InterruptType = 1;
+
+/// Data type to store an interrupt event.
+pub type InterruptEvent = u32;
+
+pub const IRQ_TRIGGERED: InterruptEvent = 0;
+
+/// Trait to manage interrupt sources for virtual device backends.
+///
```

Comment:
I would like to see more details here around how we expect this to be used. I can work on something for it.

Reply:
Currently we make use of the interrupt group object as below:
```
    /// Trigger an interrupt for a specific event source.
    pub fn trigger(&self, index: u32, flags: u32) -> Result<(), std::io::Error> {
        if self.legacy_mode {
            self.intr_group.trigger(0, flags)
        } else {
            self.intr_group.trigger(index, 0)
        }
    }

    /// Get irq eventfd array for vritio vrings.
    pub fn get_vring_irqfds(&self) -> Vec<&EventFd> {
        match self.intr_group.get_type() {
            InterruptSourceType::LegacyIrq => {
                let irqfd = self.intr_group.get_irqfd(0).unwrap();
                vec![irqfd; 1]
            }
            InterruptSourceType::MsiIrq | InterruptSourceType::VfioMsiIrq(_, _) => {
                //Skip the first irqfd which is configure change
                let intr_len = self.intr_group.len() as usize;
                let mut irqfds = Vec::with_capacity(intr_len - 1);
                for index in 1..intr_len {
                    irqfds.push(self.intr_group.get_irqfd(index as InterruptIndex).unwrap());
                }
                irqfds
            }
        }
    }
```